### PR TITLE
Upgrade Scala types to use lazy properties

### PR DESCRIPTION
**Pull request by asodja**

### CODE REVIEW

**Comment by asodja** [comment]
@bot-gradle test this

**Comment by asodja** [comment]
@bot-gradle test this

**Comment by asodja** [comment]
@bot-gradle test this

**Comment by asodja** [comment]
@bot-gradle test this

File: platforms/jvm/scala/src/integTest/groovy/org/gradle/scala/compile/UpToDateScalaCompileIntegrationTest.groovy [hunk 1]
```diff
@@ -144,7 +144,10 @@ class UpToDateScalaCompileIntegrationTest extends AbstractIntegrationSpec implem
 
             tasks.withType(ScalaCompile) {
                 scalaCompileOptions.with {
-                    additionalParameters = (additionalParameters ?: []) + "-target:8"
+                    if (!additionalParameters.isPresent()) {
+                        additionalParameters = []
+                    }
+                    additionalParameters.add("-target:8")
```
Comment (lptr) [inline platforms/jvm/scala/src/integTest/groovy/org/gradle/scala/compile/UpToDateScalaCompileIntegrationTest.groovy:150]: :x: This doesn't sound right.

Whoever sets the parameter here shouldn't need to be concerned whether the list is absent or not. The user intention here is to add a parameter, and they should be able to do so with just calling `.add()`.

I'm not sure what we need to achieve this. Shall we simply initialize this particular property to `[]` by default? Should that be the default for all collection properties? What is the point of having an absent collection property?

File: platforms/jvm/scala/src/main/java/org/gradle/api/internal/tasks/scala/MinimalScalaCompileOptions.java [hunk 2]
```diff
@@ -42,17 +42,17 @@ public class MinimalScalaCompileOptions implements Serializable {
...
+        this.deprecation = compileOptions.getDeprecation().getOrElse(true);
+        this.unchecked = compileOptions.getUnchecked().getOrElse(true);
+        this.debugLevel = compileOptions.getDebugLevel().getOrNull();
+        this.optimize = compileOptions.getOptimize().getOrElse(false);
+        this.encoding = compileOptions.getEncoding().getOrNull();
+        this.force = compileOptions.getForce().getOrElse(false);
+        this.additionalParameters = ImmutableList.copyOf(compileOptions.getAdditionalParameters().get());
+        this.listFiles = compileOptions.getListFiles().getOrElse(false);
+        this.loggingLevel = compileOptions.getLoggingLevel().getOrNull();
+        this.loggingPhases = ImmutableList.copyOf(compileOptions.getLoggingPhases().get());
```
Comment (lptr) [inline platforms/jvm/scala/src/main/java/org/gradle/api/internal/tasks/scala/MinimalScalaCompileOptions.java:55]: ❓ Why do we need to pass default values here? Don't we have conventions set for all of these?

I don't see how we can avoid introducing something like a `BooleanFlagProperty` to properly handle these use cases. We need a way to encode in the type system that these properties can never be absent.

File: platforms/jvm/scala/src/main/java/org/gradle/api/plugins/scala/ScalaBasePlugin.java [hunk 3]
```diff
@@ -289,10 +289,11 @@ private static JavaVersion computeJavaTargetCompatibilityConvention(DefaultJavaP
     }
 
     private void configureScaladoc(final Project project, final ScalaRuntime scalaRuntime) {
+        ProviderFactory providers = project.getProviders();
         project.getTasks().withType(ScalaDoc.class).configureEach(scalaDoc -> {
-            scalaDoc.getConventionMapping().map("scalaClasspath", (Callable<FileCollection>) () -> scalaRuntime.inferScalaClasspath(scalaDoc.getClasspath()));
-            scalaDoc.getConventionMapping().map("destinationDir", (Callable<File>) () -> javaPluginExtension(project).getDocsDir().dir("scaladoc").get().getAsFile());
-            scalaDoc.getConventionMapping().map("title", (Callable<String>) () -> project.getExtensions().getByType(ReportingExtension.class).getApiDocTitle());
+            scalaDoc.getScalaClasspath().convention(providers.provider(() -> scalaRuntime.inferScalaClasspath(scalaDoc.getClasspath())));
```
Comment (lptr) [inline platforms/jvm/scala/src/main/java/org/gradle/api/plugins/scala/ScalaBasePlugin.java:294]: ❓ Why do we need to wrap this in a `provider()`? I'd expect the `FileCollection` returned by `inferScalaClasspath()` should be good enough already, no? Constructing that object is not very costly, and it already provides the laziness we need here.

File: platforms/jvm/scala/src/main/java/org/gradle/api/tasks/scala/ScalaDoc.java [hunk 4]
```diff
@@ -207,9 +186,9 @@ public Property<JavaLauncher> getJavaLauncher() {
     @TaskAction
     protected void generate() {
         ScalaDocOptions options = getScalaDocOptions();
-        if (!GUtil.isTrue(options.getDocTitle())) {
-            options.setDocTitle(getTitle());
-        }
+        String docTitle = !GUtil.isTrue(options.getDocTitle().getOrNull())
+            ? getTitle().getOrNull()
+            : options.getDocTitle().getOrNull();
```
Comment (lptr) [inline platforms/jvm/scala/src/main/java/org/gradle/api/tasks/scala/ScalaDoc.java:191]: How about:

```suggestion
    String docTitle = options.getDocTitle()
            .orElse(getTitle())
            .getOrNull();
```

Yes, this would accept `""` as a valid title, but I think that's actually better behavior. That's what we are doing with the footer below anyway.

File: platforms/jvm/scala/src/test/groovy/org/gradle/api/plugins/scala/ScalaBasePluginTest.groovy [hunk 5]
```diff
@@ -137,8 +147,22 @@ class ScalaBasePluginTest extends AbstractProjectBuilderSpec {
...
+        new File(artifactDir, "${artifactId}-${version}.jar").createNewFile()
+        new File(artifactDir, "${artifactId}-${version}.pom") << """
+            <project>
+                <modelVersion>4.0.0</modelVersion>
+                <groupId>$group</groupId>
+                <artifactId>$artifactId</artifactId>
+                <version>$version</version>
+            </project>
+        """
+    }
```
Comment (lptr) [inline platforms/jvm/scala/src/test/groovy/org/gradle/api/plugins/scala/ScalaBasePluginTest.groovy:167]: ❔ Why is this needed now?

File: platforms/jvm/scala/src/main/java/org/gradle/api/tasks/scala/ScalaDoc.java (hunk 4 above)
Comment (asodja) [reply platforms/jvm/scala/src/main/java/org/gradle/api/tasks/scala/ScalaDoc.java]: Sure why not. I don't expect this to be a big problem

**Comment by asodja** [comment]
@bot-gradle test this

File: platforms/jvm/scala/src/main/java/org/gradle/api/internal/tasks/scala/MinimalScalaCompileOptions.java (hunk 2 above)
Comment (asodja) [reply platforms/jvm/scala/src/main/java/org/gradle/api/internal/tasks/scala/MinimalScalaCompileOptions.java:55]: I will leave this for now. I can just remove defaults for inputs that are not `@Optional` (or `@Internal` or `@Console`) if you want. But I generally think it's not a big problem

**Comment by asodja** [comment]
@bot-gradle test this

File: platforms/jvm/scala/src/main/java/org/gradle/api/internal/tasks/scala/MinimalScalaCompileOptions.java (hunk 2 above)
Comment (lptr) [reply platforms/jvm/scala/src/main/java/org/gradle/api/internal/tasks/scala/MinimalScalaCompileOptions.java:55]: The problem I see here is that we are specifying `true` as a default for `failOnError` here, but we also do the same in `BasicScalaOptions`. It should not need to happen in both places. What if the values are different? (Essentially, the value here is useless, as `failOnError` will never be absent. But then why do we have this `true` here?)

I'm still convinced we need a way for the type system to express that these things cannot be absent.

Comment (lptr) [reply platforms/jvm/scala/src/main/java/org/gradle/api/internal/tasks/scala/MinimalScalaCompileOptions.java:55]: I think leaving this as is for now is okay, introducing a non-nullable property type should happen on its own.

**Review by lptr — APPROVED** [review]
LGTM. Let's discuss introducing a never-absent property outside of this context.
